service/history/api/respondactivitytaskcompleted/api.go

```diff
@@ -97,13 +100,26 @@ func Invoke(
 			}
 
 			if !isRunning ||
-				ai.StartedEventId == common.EmptyEventID ||
+				(!isCompletedByID && ai.StartedEventId == common.EmptyEventID) ||
 				(token.GetScheduledEventId() != common.EmptyEventID && token.Attempt != ai.Attempt) ||
 				(token.GetVersion() != common.EmptyVersion && token.Version != ai.Version) {
 				return nil, consts.ErrActivityTaskNotFound
 			}
 
-			if _, err := mutableState.AddActivityTaskCompletedEvent(scheduledEventID, ai.StartedEventId, request); err != nil {
+			// We fabricate a started event only when the activity is not started yet and
+			// we need to force complete an activity
+			if isCompletedByID && ai.StartedEventId == common.EmptyEventID {
+				_, err := mutableState.AddActivityTaskStartedEvent(ai, scheduledEventID,
+					"",
+					req.GetCompleteRequest().GetIdentity(),
+					mutableState.GetMostRecentWorkerVersionStamp())
```

Comment:
I think this should be left empty.
I'll double check.

Reply:
Yes please pass nil.